### DISCUSSION TITLE
Adding support for SMP on AArch32

### DISCUSSION THREAD

**Discussion by ayankuma** [zephyrproject-rtos/zephyr#54829] [Q&A]
Hi,

I am trying to add support for SMP for AArch32 (specifically 'fvp_baser_aemv8r_aarch32' platform).

I see that AArch32 lacks support for context switching.
I am trying to follow the context switching code for Arm64, and z_arm64_context_switch() caught my attention.

Do we know any Arm specification that states how context switching is to be done ? For eg I see that we are saving and restoring registers x19 ... x29, but not the previous registers. Do we know the reason why ?

Is there any list of registers (specified in any Arm docs) which are to be saved and restored.

@carlocaione :- I see that you are author, can you provide some pointers, please ?

**Comment by carlocaione** [comment]
> I see that AArch32 lacks support for context switching.

That's obviously not true, without context switching you would not have preemption nor multitasking.

> Do we know any Arm specification that states how context switching is to be done ?

There is no spec for that. Every OS or architecture does that differently.

>  I see that we are saving and restoring registers x19 ... x29, but not the previous registers. Do we know the reason why ?

The caller-saved registers are saved on exception entry: in https://github.com/zephyrproject-rtos/zephyr/blob/44488f1cf661fdd55a9c9546d091edf37890d620/arch/arm64/core/vector_table.S#L27

and restored on exception exit in: https://github.com/zephyrproject-rtos/zephyr/blob/44488f1cf661fdd55a9c9546d091edf37890d620/arch/arm64/core/vector_table.S#L213

The callee-saved registers are lazily saved / restored only when a context switch does happen in: https://github.com/zephyrproject-rtos/zephyr/blob/44488f1cf661fdd55a9c9546d091edf37890d620/arch/arm64/core/switch.S#L33


> Is there any list of registers (specified in any Arm docs) which are to be saved and restored.

Again, it depends (i.e. MMU vs no-MMU, FPU vs no-FPU, ASID vs no-ASID, etc...) you can have a possible comprehensive list at https://developer.arm.com/documentation/den0024/a/The-Memory-Management-Unit/Context-switching

**Reply by ayankuma** [reply]
> That's obviously not true, without context switching you would not have preemption nor multitasking.

Yes, you are correct. What I meant was that `arch_switch()` is not implemented for AArch32. Instead `arch_swap()` is used.

While trying to understand the difference, I came across this line
https://github.com/zephyrproject-rtos/zephyr/blob/44488f1cf661fdd55a9c9546d091edf37890d620/kernel/include/kernel_arch_interface.h#L97

```
This store must be done AFTER all relevant state
 * is saved, and must include whatever memory barriers or cache
 * management code is required to be sure another CPU will see the
 * result correctly.
```

However, looking at `z_arm64_context_switch()`, I don't see any barriers being used. IIUC, we would need a `dmb` at the following https://github.com/zephyrproject-rtos/zephyr/blob/44488f1cf661fdd55a9c9546d091edf37890d620/arch/arm64/core/switch.S#L53

The reason being we would want to ensure that the stores are visible to all the other cores (ie they aren't cached or buffered before a new context is retrieved).

I am trying to implement this for arm32 and thus want to make sure I understand it correctly.

**Comment by carlocaione** [comment]
> I am trying to implement this for arm32 and thus want to make sure I understand it correctly.

Yes, this is missing. You can add it to the `switch.S` file but that's half of the story, since you also would need that in: https://github.com/zephyrproject-rtos/zephyr/blob/c9225e4365b95337560948bdb93168e777eb9e42/kernel/include/kswap.h#L33-L51

Anyway, this is proved not to a be an issue (so far) on emulation and real hardware.
